File: 程式設計：資料結構答案檔/Python/502.py
def max_subarray_sum(arr):
    if len(arr) == 0:
        return 0

    def max_crossing_sum(arr, low, mid, high):
        left_sum = float('-inf')
        sum = 0
        max_left = 0
        for i in range(mid, low - 1, -1):
            sum += arr[i]
            if sum > left_sum:
                left_sum = sum
                max_left = i

        right_sum = float('-inf')
        sum = 0
        max_right = 0
        for i in range(mid + 1, high + 1):
            sum += arr[i]
            if sum > right_sum:
                right_sum = sum
                max_right = i

        return max_left, max_right, left_sum + right_sum

    def max_subarray_sum_helper(arr, low, high):
        if low == high:
            return low, high, arr[low]

        mid = (low + high) // 2

        left_low, left_high, left_sum = max_subarray_sum_helper(arr, low, mid)
        right_low, right_high, right_sum = max_subarray_sum_helper(arr, mid + 1, high)
        cross_low, cross_high, cross_sum = max_crossing_sum(arr, low, mid, high)

        if left_sum >= right_sum and left_sum >= cross_sum:
            return left_low, left_high, left_sum
        elif right_sum >= left_sum and right_sum >= cross_sum:
            return right_low, right_high, right_sum
        else:
            return cross_low, cross_high, cross_sum

    low, high, max_sum = max_subarray_sum_helper(arr, 0, len(arr) - 1)
    return low, high, max_sum

arr = [int(i) for i in input().split()]
start, end, max_sum = max_subarray_sum(arr)
print(max_sum)
print(start)
print(end)

File: 程式設計：資料結構答案檔/C#/502.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Program
{
    class SubarrayResult
    {
        public int Start { get; set; }
        public int End { get; set; }
    }

    static void Main()
    {
        var arr = new List<int>();
        var input = Console.ReadLine().Split(' ');
        foreach (var item in input)
        {
            arr.Add(int.Parse(item));
        }

        if (arr.Count == 0)
        {
            Console.WriteLine("0\n0\n0");
            return;
        }

        var result = MaxSubarraySum(arr);
        int maxSum = 0;
        for (int i = result.Start; i <= result.End; i++)
        {
            maxSum += arr[i];
        }

        Console.WriteLine(maxSum);
        Console.WriteLine(result.Start);
        Console.WriteLine(result.End);
    }

    static SubarrayResult MaxSubarraySum(List<int> arr)
    {
        return MaxSubarraySumHelper(arr, 0, arr.Count - 1);
    }

    static SubarrayResult MaxSubarraySumHelper(List<int> arr, int low, int high)
    {
        if (low == high)
        {
            return new SubarrayResult { Start = low, End = high };
        }

        int mid = (low + high) / 2;
        var left = MaxSubarraySumHelper(arr, low, mid);
        var right = MaxSubarraySumHelper(arr, mid + 1, high);
        var crossing = MaxCrossingSum(arr, low, mid, high);

        int leftSum = arr.GetRange(left.Start, left.End - left.Start + 1).Sum();
        int rightSum = arr.GetRange(right.Start, right.End - right.Start + 1).Sum();
        int crossingSum = arr.GetRange(crossing.Start, crossing.End - crossing.Start + 1).Sum();

        if (leftSum >= rightSum && leftSum >= crossingSum)
        {
            return left;
        }
        else if (rightSum >= leftSum && rightSum >= crossingSum)
        {
            return right;
        }
        else
        {
            return crossing;
        }
    }

    static SubarrayResult MaxCrossingSum(List<int> arr, int low, int mid, int high)
    {
        int leftSum = int.MinValue, rightSum = int.MinValue;
        int sum = 0;
        int maxLeft = mid, maxRight = mid;

        for (int i = mid; i >= low; i--)
        {
            sum += arr[i];
            if (sum > leftSum)
            {
                leftSum = sum;
                maxLeft = i;
            }
        }

        sum = 0;
        for (int i = mid + 1; i <= high; i++)
        {
            sum += arr[i];
            if (sum > rightSum)
            {
                rightSum = sum;
                maxRight = i;
            }
        }

        return new SubarrayResult { Start = maxLeft, End = maxRight };
    }
}
